src/supervised/ordinal.jl

```diff
@@ -0,0 +1,49 @@
+"""
+    OrdinalMarginLoss <: SupervisedLoss
+
+Modifies a margin loss `loss` to be used on an ordinal domain with
+number of levels `nlevels`. It treats each level as an integer between
+1 and `nlevels`, inclusive, and penalizes output according to the sum of
+level thresholds crossed relative to target
+
+Assumes target is encoded in an Index encoding scheme where levels are
+numbered between 1 and `nlevels`
+"""
+struct OrdinalMarginLoss <: SupervisedLoss
+    loss::MarginLoss
+    nlevels::Int
+end
+
+for fun in (:value, :deriv, :deriv2)
+    @eval @fastmath function ($fun)(loss::OrdinalMarginLoss, target::Number, output::Number)
+        retval = 0
```

Comment:
Thats because in

```julia
struct OrdinalMarginLoss <: SupervisedLoss
    loss::MarginLoss
    nlevels::Int
end
```

the member `loss`  is weakly typed. This means no matter how beautifully clean the function `value`, the call `value(loss.loss,...)` will always be a run-time look up where the compiler can't know which method will be called . Try defining the type as I outlined in an earlier post using type variables